doc/feature/cppPlugins.md

```diff
@@ -0,0 +1,175 @@
+C++ Plugin API for CPU compressors
+============
+
+This document specifies a modern API for the in-memory C plugin API. It
+will not replace the CPU-GPU transfer plugins.
+
+## Requirements
+
+* Support all features used by current CPU compression plugins
+    * Support 1D and 2D data layouts
+    * Drop alpha during compression of 2D data with alpha
+    * Lossy compression for 2D data
+    * Support for common image pixel layouts
+    * Support for different 1D data (byte, float, double, ...)
+    * Multiple compressors per plugin DSO
+* Information provided for each compressor:
+    * Compression speed (relative to RLE compressor)
+    * Compression ratio (relative to RLE compressor)
+    * 1D or 2D compressor
+    * Implemention drops alpha (for 2D)
+* Compressors may produce multiple output chunks
+* Compressor engines are instantiated to allow reuse of internal state
+* Easy to use API for plugin implementers
+
+## Dependency Changes
+
+* New: Lunchbox plugin API for loading plugins
+
+## API
+
+Change signature of Lunchbox::Plugin::HandlesFunc to:
+    typedef boost::function< bool ( InitDataT& ) > HandlesFunc;
```

Comment:
Here's what I propose, it's sketchy but I think it provides what you need without breaking the API or adding debatable semantics to the existing functions.
I don't like the fact that is using enable_if because it makes it ugly, but hopefully nobody needs to look at the code itself, but at a clear and concise example.

In real file, the PluginRegisterer class needs to be also extended to support the registration of plugins with traits, but maybe that's optional.
I haven't thought about how to combine this with dynamic library loading, I want to know first if you think this approach is better.

```
#include <functional>
#include <memory>
#include <type_traits>
#include <string>
#include <vector>

#include <boost/functional/factory.hpp>

namespace servus
{
    typedef std::string URI;
    typedef int uint128_t;
    uint128_t make_UUID() { return 0; }
}

template< class PluginT, class InitDataT = servus::URI, class TraitsT = void >
class PluginFactory;

template< class PluginT, class InitDataT = servus::URI, class TraitsT = void >
class Plugin
{
public:
    typedef std::function< PluginT*( const InitDataT& )> Constructor;
    typedef std::function< bool( const InitDataT& )> HandlesFunc;

    const static bool hasTraits = !std::is_same< TraitsT, void >::value;

    // This constructor requires TraitsT to be default constructible otherwise
    // we have to disable it if TraitsT is not void
    Plugin( const Constructor& constructor_, const HandlesFunc& handles_ )
        : constructor( constructor_ ), handles( handles_ )
        , tag( servus::make_UUID( )) {}

    template< typename T = TraitsT >
    Plugin( const Constructor& constructor_, const HandlesFunc& handles_,
            const typename std::enable_if< hasTraits, T >::type& traits_ )
        : constructor( constructor_ ), handles( handles_ )
        , tag( servus::make_UUID( ))
        , traits( traits_ )
    {}

    bool operator == ( const Plugin& rhs ) const
        { return tag == rhs.tag; }

    bool operator != ( const Plugin& rhs ) const { return !(*this == rhs); }

    template< typename T = TraitsT >
    const T& getTraits(
        typename std::enable_if< hasTraits, T >::type* = 0) const
    {
        return traits;
    }

private:
    friend class PluginFactory< PluginT, InitDataT, TraitsT >;
    Constructor constructor;
    HandlesFunc handles;
    servus::uint128_t tag;
    // We could also use std::unique_ptr<TraitsT>, but then the default copy
    // constructor gets deleted and we have to write one.
    typename std::conditional< hasTraits, TraitsT, void* >::type traits;
};


template< class PluginT, class InitDataT, class TraitsT >
class PluginFactory
{
public:
    typedef Plugin< PluginT, InitDataT, TraitsT > PluginHolder;
    typedef std::vector< PluginHolder > Plugins;

    typedef std::function< size_t( std::vector< TraitsT >)> ChooseFunc;

    PluginT* create( const InitDataT& initData )
    {
        return new PluginT( initData );
    }


    template< typename T = PluginT >
    T* create( const InitDataT& initData, const ChooseFunc& chooser ) const
    {
        std::vector< TraitsT > traits;
        Plugins candidates;
        for( auto& plugin : _plugins )
        {
            if( plugin.handles( initData ))
            {
                candidates.push_back( plugin );
                traits.push_back( plugin.getTraits( ));
            }
        }

        if( candidates.empty( ))
            return nullptr;

        return candidates[chooser( traits )].constructor( initData );
    }

    void register_( const Plugin< PluginT, InitDataT, TraitsT >& plugin )
    {
        _plugins.push_back( plugin );
    }

private:
    Plugins _plugins;
};


class MyPlugin
{
public:
    MyPlugin( const servus::URI& uri )
    {
    }

    static bool handles( const servus::URI& uri )
    {
        return true;
    }
};

class MyTraits
{
};

typedef PluginFactory< MyPlugin, servus::URI, MyTraits > FactoryWithTraits;
typedef PluginFactory< MyPlugin, servus::URI > FactoryWithoutTraits;

FactoryWithTraits factoryWithTraits;
FactoryWithoutTraits factoryWithoutTraits;

size_t chooser( const std::vector< MyTraits >& traits )
{
    return 0;
}

int main()
{
    factoryWithTraits.register_(
        Plugin< MyPlugin, servus::URI, MyTraits >(
            std::bind( boost::factory< MyPlugin* >(), std::placeholders::_1 ),
            std::bind( &MyPlugin::handles, std::placeholders::_1 ),
            MyTraits( )));

    factoryWithoutTraits.register_(
        Plugin< MyPlugin, servus::URI >(
            std::bind( boost::factory< MyPlugin* >(), std::placeholders::_1 ),
            std::bind( &MyPlugin::handles, std::placeholders::_1 )));

    MyPlugin* plugin = factoryWithTraits.create( "initData", chooser );
    MyPlugin* plugin2 = factoryWithoutTraits.create( "initData" );
}
```